File: apps/ios/Sources/Parsing.swift
import Foundation

enum Parsing {
    static func parseExampleResponse(_ text: String) -> ExampleOutput? {
        let normalized = text.replacingOccurrences(of: "\\n", with: "\n").trimmingCharacters(in: .whitespacesAndNewlines)

        func lineValue(_ label: String) -> String {
            let pattern = "^\\s*" + NSRegularExpression.escapedPattern(for: label) + ":\\s*(.+)$"
            let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive])
            for line in normalized.components(separatedBy: "\n") {
                if let match = regex?.firstMatch(in: line, options: [], range: NSRange(location: 0, length: line.utf16.count)) {
                    if let range = Range(match.range(at: 1), in: line) {
                        return String(line[range]).trimmingCharacters(in: .whitespacesAndNewlines)
                    }
                }
            }
            return ""
        }

        func blockValue(_ label: String) -> String {
            let pattern = NSRegularExpression.escapedPattern(for: label) + ":\\s*([\\s\\S]+)$"
            let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive])
            if let match = regex?.firstMatch(in: normalized, options: [], range: NSRange(location: 0, length: normalized.utf16.count)) {
                if let range = Range(match.range(at: 1), in: normalized) {
                    return String(normalized[range]).trimmingCharacters(in: .whitespacesAndNewlines)
                }
            }
            return ""
        }

        let jp = lineValue("JP")
        let reading = lineValue("Reading")
        let zh = lineValue("ZH")
        let grammar = blockValue("Grammar")

        guard !jp.isEmpty, !reading.isEmpty, !zh.isEmpty, !grammar.isEmpty else { return nil }
        return ExampleOutput(jp: jp, reading: reading, zh: zh, grammar: grammar)
    }

    static func normalizeTranslation(_ raw: String) -> String? {
        let normalized = raw.replacingOccurrences(of: "\\n", with: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
        let firstLine = normalized
            .components(separatedBy: "\n")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .first { !$0.isEmpty }

        guard var line = firstLine else { return nil }
        line = line.replacingOccurrences(of: "^zh[:：]\\s*", with: "", options: [.regularExpression, .caseInsensitive])
        line = line.replacingOccurrences(of: "^translation[:：]\\s*", with: "", options: [.regularExpression, .caseInsensitive])
        line = line.trimmingCharacters(in: .whitespacesAndNewlines)

        line = line.replacingOccurrences(of: "^['\"「『](.*)['\"」』]$", with: "$1", options: [.regularExpression])
        let result = line.trimmingCharacters(in: .whitespacesAndNewlines)
        return result.isEmpty ? nil : result
    }

    static func parseChoiceResponse(_ text: String) -> [String] {
        let normalized = text.replacingOccurrences(of: "\\n", with: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty else { return [] }
        return normalized
            .components(separatedBy: "\n")
            .map { line in
                line.replacingOccurrences(of: "^[\\s*\\d.\\)\\(-]+", with: "", options: [.regularExpression])
                    .trimmingCharacters(in: .whitespacesAndNewlines)
            }
            .filter { !$0.isEmpty }
    }
}

File: apps/android/app/src/main/java/com/learnjapanese/app/data/Parsing.kt
package com.learnjapanese.app.data

object Parsing {
    private fun cleanModelText(input: String): String {
        return input
            .replace(Regex("\\*\\*(.*?)\\*\\*"), "$1")
            .replace(Regex("__(.*?)__"), "$1")
            .replace(Regex("`(.*?)`"), "$1")
            .replace(Regex("^[-*]\\s+"), "")
            .trim()
    }

    fun parseExampleResponse(text: String): ExampleOutput? {
        val normalized = text.replace("\\n", "\n").trim()

        fun lineValue(label: String): String {
            val regex =
                Regex(
                    "^\\s*(?:\\*\\*\\s*)?" + Regex.escape(label) + "\\s*(?:\\*\\*\\s*)?:\\s*(.+)$",
                    RegexOption.IGNORE_CASE,
                )
            return normalized
                .lineSequence()
                .mapNotNull { line ->
                    regex
                        .find(line)
                        ?.groupValues
                        ?.getOrNull(1)
                        ?.let(::cleanModelText)
                }.firstOrNull()
                .orEmpty()
        }

        fun blockValue(label: String): String {
            val regex =
                Regex(
                    "(?:\\*\\*\\s*)?" + Regex.escape(label) + "\\s*(?:\\*\\*\\s*)?:\\s*([\\s\\S]+)$",
                    RegexOption.IGNORE_CASE,
                )
            return regex
                .find(normalized)
                ?.groupValues
                ?.getOrNull(1)
                ?.let(::cleanModelText)
                .orEmpty()
        }

        val jp = lineValue("JP")
        val reading = lineValue("Reading")
        val zh = lineValue("ZH")
        val grammar = blockValue("Grammar")

        return if (jp.isNotEmpty() && reading.isNotEmpty() && zh.isNotEmpty() && grammar.isNotEmpty()) {
            ExampleOutput(jp = jp, reading = reading, zh = zh, grammar = grammar)
        } else {
            null
        }
    }

    fun normalizeTranslation(raw: String): String? {
        val normalized = raw.replace("\\n", "\n").trim()
        val firstLine =
            normalized
                .lines()
                .map { it.trim() }
                .firstOrNull { it.isNotEmpty() }
                ?: return null

        var line = firstLine
        line = line.replace(Regex("^zh[:：]\\s*", RegexOption.IGNORE_CASE), "")
        line = line.replace(Regex("^translation[:：]\\s*", RegexOption.IGNORE_CASE), "")
        line = line.trim()
        line = line.replace(Regex("^['\"「『](.*)['\"」』]$"), "$1").trim()
        line = cleanModelText(line)
        return line.ifEmpty { null }
    }

    fun parseChoiceResponse(text: String): List<String> {
        val normalized = text.replace("\\n", "\n").trim()
        if (normalized.isEmpty()) return emptyList()
        return normalized
            .lines()
            .map { line ->
                line.replace(Regex("^[\\s*\\d.\\)\\(-]+"), "").trim()
            }.filter { it.isNotEmpty() }
    }
}
